File: benchmarks/complex/pipeline.cpp
#include <cstdio>
#include <vector>
using std::vector;
static void qsort_(vector<long>& a, long lo, long hi) {
    if (lo < hi) {
        long p = a[hi], i = lo - 1;
        for (long j = lo; j < hi; j++) if (a[j] < p) { i++; long t = a[i]; a[i] = a[j]; a[j] = t; }
        long t = a[i + 1]; a[i + 1] = a[hi]; a[hi] = t;
        qsort_(a, lo, i); qsort_(a, i + 2, hi);
    }
}
static long bsearch_(const vector<long>& a, long n, long key) {
    long lo = 0, hi = n - 1;
    while (lo <= hi) { long mid = (lo + hi) / 2; if (a[mid] == key) return mid; if (a[mid] < key) lo = mid + 1; else hi = mid - 1; }
    return -1;
}
int main() {
    long n = 200000; vector<long> a(n); long seed = 12345;
    for (long i = 0; i < n; i++) { seed = (seed * 1103515245 + 12345) % 2147483648L; a[i] = seed % 1000000; }
    qsort_(a, 0, n - 1);
    long hits = 0;
    for (long q = 0; q < 20000; q++) if (bsearch_(a, n, q * 50) >= 0) hits++;
    long hist[256] = {0};
    for (long i = 0; i < n; i++) hist[a[i] % 256]++;
    long checksum = 0;
    for (long k = 0; k < 256; k++) checksum = (checksum + hist[k] * (k + 1)) % 1000000007L;
    printf("%ld\n", hits * 1000000007L % 1000000007L + checksum * 1000 + hits);
}

File: benchmarks/complex/pipeline.rs
fn qsort(a: &mut [i64], lo: i64, hi: i64) {
    if lo < hi {
        let p = a[hi as usize];
        let mut i = lo - 1;
        let mut j = lo;
        while j < hi { if a[j as usize] < p { i += 1; a.swap(i as usize, j as usize); } j += 1; }
        a.swap((i + 1) as usize, hi as usize);
        qsort(a, lo, i);
        qsort(a, i + 2, hi);
    }
}
fn bsearch(a: &[i64], n: i64, key: i64) -> i64 {
    let (mut lo, mut hi) = (0i64, n - 1);
    while lo <= hi {
        let mid = (lo + hi) / 2;
        if a[mid as usize] == key { return mid; }
        if a[mid as usize] < key { lo = mid + 1; } else { hi = mid - 1; }
    }
    -1
}
fn main() {
    let n = 200000i64;
    let mut a = vec![0i64; n as usize];
    let mut seed = 12345i64;
    for i in 0..n as usize { seed = (seed * 1103515245 + 12345) % 2147483648; a[i] = seed % 1000000; }
    qsort(&mut a, 0, n - 1);
    let mut hits = 0i64;
    for q in 0..20000i64 { if bsearch(&a, n, q * 50) >= 0 { hits += 1; } }
    let mut hist = [0i64; 256];
    for i in 0..n as usize { let b = (a[i] % 256) as usize; hist[b] += 1; }
    let mut checksum = 0i64;
    for k in 0..256i64 { checksum = (checksum + hist[k as usize] * (k + 1)) % 1000000007; }
    println!("{}", hits * 1000000007 % 1000000007 + checksum * 1000 + hits);
}
